### DISCUSSION TITLE
EvaluateScriptAsync for SelfContained application

### DISCUSSION THREAD

**Discussion by Zelya** [cefsharp/CefSharp#4728] [General]
### Is there an existing issue for this?

- [X] I have searched both open/closed issues, no issue already exists.

### CefSharp Version

121.3.70

### Operating System

Windows 11

### Architecture

x64

### .Net Version

NET 8.0

### Implementation

WPF

### Reproduction Steps

1. Add EvaluateScriptAsync code to your project, like this

```
        private async void Browser_LoadingStateChanged(object? sender, CefSharp.LoadingStateChangedEventArgs e)
        {
            string script = @"document.querySelector('someelement')";

            var response = await Browser.GetFocusedFrame().EvaluateScriptAsync(script, null);

            if (response.Success)
            {
```
It works OK.
2.  In order to make browser working with SelfContained=true (see  https://github.com/cefsharp/CefSharp/issues/3407) add code sninppet from MinimalExample https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/CefSharp.MinimalExample.Wpf/Program.PublishSingleFile.cs
This is the most valuable part. 
```
            var settings = new CefSettings()
            {
                CachePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "CefSharp\\Cache"),
                BrowserSubprocessPath = System.Diagnostics.Process.GetCurrentProcess().MainModule.FileName
            };

            Cef.Initialize(settings, performDependencyCheck: false);
```
3. Now EvaluateScriptAsync hangs forever.

### Expected behavior

EvaluateScriptAsync should work for app with SelfContained=true

### Actual behavior

EvaluateScriptAsync hangs for app with SelfContained=true and workaround from MinimalExample

### Does this problem also occur in the CEF Sample Application

Not Tested

**Comment by amaitland** [comment]
Follow the instructions at https://github.com/cefsharp/CefSharp/wiki/SelfHost-BrowserSubProcess

Make sure to change the stack size of your application.

> private async void Browser_LoadingStateChanged(object? sender, CefSharp.LoadingStateChangedEventArgs e)
>         {
>             string script = @"document.querySelector('someelement')";
> 
>             var response = await Browser.GetFocusedFrame().EvaluateScriptAsync(script, null);
> 
>             if (response.Success)
>             {

There are two problems with this code block

1. You aren't checking if the browser is actually loaded. e.g. `LoadingStateChangedEventArgs.IsLoading`
2. Returning `DOM` elements directly is not supported.

> You cannot directly return a DOM [Element](https://developer.mozilla.org/en-US/docs/Web/API/Element) (or anything with a cyclic reference), you need to create a new object that has just the information you need to return.

As per https://github.com/cefsharp/CefSharp/wiki/General-Usage#2-how-do-you-call-a-javascript-method-that-returns-a-result

For simple elements it might work with an 8mb stack size. It's likely crashing the render process in your case.

See https://github.com/cefsharp/CefSharp.Dom for an alternative DOM API.

**Reply by Zelya** [reply]
@amaitland 
> You aren't checking if the browser is actually loaded. e.g. LoadingStateChangedEventArgs.IsLoading

Sorry, skipped this for smaller code snippet. But this doesn't help.

> Returning DOM elements directly is not supported.

The same. Code like `string script = @"document.querySelector('someelement').value;"` doesn't work too.

> See https://github.com/cefsharp/CefSharp.Dom for an alternative DOM API.

Tried, but it hangs on `await chromiumWebBrowser.CreateDevToolsContextAsync();`

Without SelfHost BrowserSubProcess everything works good.

**Reply by amaitland** [reply]
Quick test with https://github.com/cefsharp/CefSharp.MinimalExample changing the netcore WPF Example to SelfContained and targeting .Net 6.0. `EvaluateScriptAsync` is working as expected

I'd suggest testing with the MinimalExample and seeing how it behaves. For `SelfContained` builds that target `net6.0-windows` it will default to self hosting.

> Follow the instructions at https://github.com/cefsharp/CefSharp/wiki/SelfHost-BrowserSubProcess

Make sure you double check you've implemented everything as per the guide. This is more up to date than the issue you linked.
